### PR TITLE
Madelynolivia/airtable integration

### PR DESCRIPTION
**Pull request by madelyndubuk**
## Ticket Information

**Link to Notion**
https://www.notion.so/ctc/Airtable-Api-Client-581c8a933cc24b51becc7d2855b7fdd5
**Short Description**
We created a wrapper around the AirTable client.  The function we support lists all of the organizations with optional filters.
## PR Checklist

- [x] Wrote ava tests
- [ ] Reviewed by **all** @ctcusc/RiverLA team members

### CODE REVIEW

File: src/apiClients/AirTableApiClient.ts [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
 class AirTableApiClient {
   private apiKey: string;
+  public base: any;
 
   constructor(apiKey: string) {
     this.apiKey = apiKey;
+    this.base = new Airtable({ apiKey: apiKey }).base('appEHr8iHguvEfXTQ');
     // remove this log when this API key is finally used somewhere
-    console.log(`SendGridAPIClient initialized with api key ${this.apiKey}`);
+    console.log(`AirTableAPIClient initialized with api key ${this.apiKey}`);
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: We can remove this log and its comment now

File: src/apiClients/AirTableApiClient.ts [hunk 2]
```diff
@@ -24,9 +35,61 @@ class AirTableApiClient {
...
+    let organizations: Organization[] = [];
+
+    const promise: Promise<Organization[]> = new Promise((resolve, reject) => {
+      this.base('Organizations')
+        .select({
+          view: 'Grid view',
+        })
+        .eachPage(
+          (records: any[], fetchNextPage: any) => {
+            let filteredRecords: any[] = records;
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: I think it'd be better to map the records into organizations immediately, so we don't have to deal with AirTable's `.get` function throughout. This will also help TypeScript do a better job here. Here's some code to get your started.

```ts

type TFieldName =
  | 'Name'
  | 'Description'
  | 'Activity'
  | 'Interest Categories'
  | 'River Section'
  | 'Phone Number'
  | 'URL'
  | 'Email';

interface Record {
  get: (fieldName: TFieldName) => string | string[];
}

function recordToOrganization(record: Record): Organization {
  return {
    name: record.get('Name') as string,
    description: record.get('Description') as string,
    activity: record.get('Activity') as string,
    interestCategories: record.get('Interest Categories') as string[],
    riverSection: record.get('River Section') as string,
    phoneNumber: record.get('Phone Number') as string,
    url: record.get('URL') as string,
    email: record.get('Email') as string,
  };
}
// your function in .eachPage should look something like this now
const func = (records: Record[], fetchNextPage: Function) => {
  let filteredOrganizations = records.map(recordToOrganization);
  // modify the rest
};

```

File: src/apiClients/AirTableApiClient.ts [hunk 3]
```diff
@@ -24,9 +35,61 @@ class AirTableApiClient {
...
+              filteredRecords = records.filter(record => {
+                return riverSections.includes(record.get('River Section'));
+              });
+            }
+            if (typeof interestCategories !== 'undefined') {
+              filteredRecords = filteredRecords.filter(record => {
+                return (
+                  interestCategories.filter(category => {
+                    return record.get('Interest Categories').includes(category);
+                  }).length != 0
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: Use `.some`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

File: src/apiClients/AirTableApiClient.ts [hunk 4]
```diff
@@ -24,9 +35,61 @@ class AirTableApiClient {
...
+                  phoneNumber: record.get('Phone Number'),
+                  url: record.get('URL'),
+                  email: record.get('Email'),
+                };
+                return org;
+              }),
+            );
+            fetchNextPage();
+          },
+          function done(err: any) {
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: I believe there's a javascript `Error` type that you can use here

File: src/apiClients/AirTableApiClient.ts [hunk 5]
```diff
@@ -1,20 +1,31 @@
...
 
   constructor(apiKey: string) {
     this.apiKey = apiKey;
+    this.base = new Airtable({ apiKey: apiKey }).base('appEHr8iHguvEfXTQ');
     // remove this log when this API key is finally used somewhere
-    console.log(`SendGridAPIClient initialized with api key ${this.apiKey}`);
+    console.log(`AirTableAPIClient initialized with api key ${this.apiKey}`);
   }
 
   async getAllOrganizations(filters: AirTableFilters = {}): Promise<Organization[]> {
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: Since we're just returning a promise, we can remove `async`

```suggestion
  getAllOrganizations(filters: AirTableFilters = {}): Promise<Organization[]> {
```

File: src/apiClients/AirTableApiClient.ts [hunk 6]
```diff
@@ -24,9 +35,61 @@ class AirTableApiClient {
...
+                  interestCategories.filter(category => {
+                    return record.get('Interest Categories').includes(category);
+                  }).length != 0
+                );
+              });
+            }
+            organizations = organizations.concat(
+              filteredRecords.map((record: any) => {
+                if (typeof record === 'undefined') {
+                }
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: why is this here? (although this entire `.map` call should disappear if you implement the advice above)

File: src/apiClients/AirTableApiClient.ts [hunk 7]
```diff
@@ -24,9 +35,61 @@ class AirTableApiClient {
...
+                const org: Organization = {
+                  name: record.get('Name'),
+                  description: record.get('Description'),
+                  activity: record.get('Activity'),
+                  interestCategories: record.get('Interest Categories'),
+                  riverSection: record.get('River Section'),
+                  phoneNumber: record.get('Phone Number'),
+                  url: record.get('URL'),
+                  email: record.get('Email'),
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: How fragile is this? Does it break when fields are empty?

File: test/apiClients/AirtableApiClient.ts [hunk 8]
```diff
@@ -1,5 +1,74 @@
 import test from 'ava';
+import sinon from 'sinon';
+import AirTableApiClient, { Organization } from '../../src/apiClients/AirTableApiClient';
+//import env from '../../src/env';
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: remove comment

File: test/apiClients/AirtableApiClient.ts [hunk 9]
```diff
@@ -1,5 +1,74 @@
...
 
-test('Returns one page of results properly', t => {
-  t.true(true, 'replace this test');
+test('Returns one page of results properly', async t => {
+  const airtableApiClient: AirTableApiClient = new AirTableApiClient('keyzxy');
+  const eachPageStub = sinon.stub().returns(records);
+  sinon.stub(airtableApiClient.base, 'select').returns({ eachPage: eachPageStub });
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: I thought we were using `nock` here. This test isn't very meaningful if we just stub the entire `eachPage` function, since it doesn't test your callback

**Comment by QuinnNgo** [comment]
@madelyndubuk @olivia-hong If you did any manual testing to make sure this works, would you mind writing details about it under the "Manual Test Plan" section of your PR? Otherwise, just fill it in with "N/A"

File: src/apiClients/AirTableApiClient.ts (hunk 7 above)
Comment (olivia-hong) [reply src/apiClients/AirTableApiClient.ts]: If any fields are empty, we just won't return that record

File: src/apiClients/AirTableApiClient.ts (hunk 2 above)
Comment (olivia-hong) [reply src/apiClients/AirTableApiClient.ts]: First filter for undefined records, then map all records into Organization objects, then filter by interesteCategory/riverSection

Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: I sent @olivia-hong a types file that should allow us to get rid of all of these `any` times. Ideally, everything would be typed so we can take full advantage of TypeScript's features.

File: test/apiClients/AirtableApiClient.ts [hunk 10]
```diff
@@ -1,5 +1,223 @@
...
+  email: 'fakeemail@chrysalis.org',
+  phoneNumber: '(999) 999-9999',
+  url: 'https://changelives.org/',
+  interestCategories: ['Social Justice and Recreation'],
+};
+
+test('Returns one page of results properly', async t => {
+  const airtableApiClient: AirTableApiClient = new AirTableApiClient('keyzxy');
+
+  nock('https://api.airtable.com/v0/appEHr8iHguvEfXTQ')
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: Can we pull this `https://api.airtable.com/v0/appEHr8iHguvEfXTQ` up into an `AIRTABLE_API_URL` variable at the top of the file?

File: test/apiClients/AirtableApiClient.ts [hunk 11]
```diff
@@ -1,5 +1,223 @@
...
+    .persist()
+    .log(data => console.log(data))
+    .get('/Organizations?view=Grid+view')
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: Use this style everywhere to help readability
```suggestion
    .get('/Organizations')
    .query({ view: 'Grid view' })
```

File: test/apiClients/AirtableApiClient.ts [hunk 12]
```diff
@@ -1,5 +1,223 @@
...
+  url: 'https://changelives.org/',
+  interestCategories: ['Social Justice and Recreation'],
+};
+
+test('Returns one page of results properly', async t => {
+  const airtableApiClient: AirTableApiClient = new AirTableApiClient('keyzxy');
+
+  nock('https://api.airtable.com/v0/appEHr8iHguvEfXTQ')
+    .persist()
+    .log(data => console.log(data))
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: Let's remove these logs

File: src/apiClients/AirTableApiClient.ts (hunk 2 above)
Comment (QuinnNgo) [reply src/apiClients/AirTableApiClient.ts]: On second thought @olivia-hong @madelyndubuk, typing this is probably going to be pretty challenging and blow out of scope -- I'd still like to do that work, but we'll tackle it in a future ticket.

File: src/apiClients/AirTableApiClient.ts [hunk 13]
```diff
@@ -1,20 +1,31 @@
...
+  interestCategories: string[];
+  riverSection: string;
+  phoneNumber: string;
+  url: string;
+  email: string;
 }
 
 class AirTableApiClient {
   private apiKey: string;
+  public base: any;
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: Does this need to be public?

File: test/apiClients/AirtableApiClient.ts [hunk 14]
```diff
@@ -1,5 +1,223 @@
...
+  const airTableFilters: AirTableFilters = {
+    interestCategories: ['Environmental Causes'],
+  };
+
+  const organizations: Organization[] = await airtableApiClient.getAllOrganizations(airTableFilters);
+  const answer: Organization[] = [org2];
+  t.deepEqual(organizations, answer);
+});
+
+test('Returns one page of filtered both empty results properly', async t => {
```
Comment (QuinnNgo) [inline test/apiClients/AirtableApiClient.ts]: Run each of these tests with `test.serial`. Ava's tests run in parallel by default, and this can mess with mocking.
```suggestion
test.serial('Returns one page of filtered both empty results properly', async t => {
```

File: src/apiClients/AirTableApiClient.ts [hunk 15]
```diff
@@ -1,32 +1,107 @@
...
+                return riverSections.includes(organization.riverSection);
+              });
+            }
+            if (typeof interestCategories !== 'undefined') {
+              organizations = organizations.filter(organization => {
+                return interestCategories
+                  .filter(category => {
+                    return organization.interestCategories.includes(category);
+                  })
+                  .some(empty);
```
Comment (QuinnNgo) [inline src/apiClients/AirTableApiClient.ts]: Whoops! Sorry I wasn't clear. This is meant to be:

```js
return interestCategories.some(category => organization.interestCategories.includes(category));
```
